pandas/_libs/groupby_helper.pxi.in

```diff
@@ -444,6 +444,116 @@ def group_nth_{{name}}(ndarray[{{dest_type2}}, ndim=2] out,
                 else:
                     out[i, j] = resx[i, j]
 
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def group_rank_{{name}}(ndarray[float64_t, ndim=2] out,
+                        ndarray[{{c_type}}, ndim=2] values,
+                        ndarray[int64_t] labels,
+                        bint is_datetimelike, **kwargs):
+    """
+    Only transforms on axis=0
+    """
+    cdef:
+        int tiebreak
+        Py_ssize_t i, j, N, K
+        int64_t val_start=0, grp_start=0, dups=0, sum_ranks=0, grp_vals_seen=1
+        int64_t grp_na_count=0
+        ndarray[int64_t] _as
+        ndarray[{{c_type}}] masked_vals
+        ndarray[uint8_t] mask
+        bint pct, ascending, keep_na
+        {{c_type}} nan_fill_val
+
+    tiebreak = tiebreakers[kwargs['ties_method']]
+    ascending = kwargs['ascending']
+    pct = kwargs['pct']
+    keep_na = kwargs['na_option'] == 'keep'
+    N, K = (<object> values).shape
+
+    masked_vals = np.array(values[:, 0], copy=True)
+    {{if name=='int64'}}
+    mask = (masked_vals == {{nan_val}}).astype(np.uint8)
+    {{else}}
+    mask = np.isnan(masked_vals).astype(np.uint8)
+    {{endif}}
+
+    if ascending ^ (kwargs['na_option'] == 'top'):
+        {{if name == 'int64'}}
+        nan_fill_val = np.iinfo(np.int64).max
+        {{else}}
+        nan_fill_val = np.inf
+        {{endif}}
+        order = (masked_vals, mask, labels)
+    else:
+        {{if name == 'int64'}}
+        nan_fill_val = np.iinfo(np.int64).min
+        {{else}}
+        nan_fill_val = -np.inf
+        {{endif}}
+        order = (masked_vals, ~mask, labels)
+    np.putmask(masked_vals, mask, nan_fill_val)
+
+    _as = np.lexsort(order)
+
+    if not ascending:
+        _as = _as[::-1]
+
+    with nogil:
+        for i in range(N):
+            dups += 1
+            sum_ranks += i - grp_start + 1
+
+            if keep_na and masked_vals[_as[i]] == nan_fill_val:
```

Comment:
This comparison to the `nan_fill_val` gets a little tricky because it obfuscates `np.nan` with `np.inf` (or whatever fill value is being used). That said, that is a pre-existing limitation which I opened #19538 to address